server.py

```diff
@@ -1,90 +1,385 @@
 
+        linking_info[text_mention] = [mid, full_mention, candidates, summary]
+    return linking_info
 
 
-@app.route('/', methods = ['GET'])
-def homePage():
-    print render_template( 'ner-home.html' )
-    return render_template( 'ner-home.html' )
+def retrieve_candidate_mids(mention):
+    """
+    Retrieve the candidate MIDs for the given mention. They are located in CANDIDATE_FILE in a certain format.
+    :param mention: The mention
+    :type mention: str
+    :return: The candidate MIDs
+    :rtype: array of strings
+    """
+    file = open(CANDIDATE_FILE, "r")
+    candidates = []
+    for line in file:
+        info = line.split("\t")
+        text_mention = info[1]
+        if mention == text_mention:
+            candidates = info[5].strip('\n')
+            candidates = candidates.split(" || ")
+            candidates = [c.strip("[") for c in candidates]
+            candidates = [c.strip("]") for c in candidates]
+    return candidates
 
 
-@app.route('/', methods = ['POST'] )
+@app.route('/', methods=['GET'])
+def home_page():
+    """
+    Renders the home page.
+    """
+    print(render_template(u"ner-home.html"))
+    return render_template(u"ner-home.html")
+
+
+@app.route('/', methods=['POST'])
 def annotate():
+    """
+    Responds to the ajax request fired when user submits the text to be detected.
+    Returns a JSON object: {'text': text, 'entities': entity info, 'lang': language of the text, 'notes': error notes}
+    """
+    start_time = time.time()
     mode = request.form['mode']
+    text = request.form['text'].strip()
+    selected = request.form['lang']
+    notes = ""
+
+    sentence = text
+
+    language = "eng"
+    if selected == "Spanish":
+        language = "spa"
+
+    elif selected == "Chinese":
+        language = "cmn"
+
+    # -------------------------- Language detector ---------------------------------
+
+    lang_detect = detect(text)
+    if lang_detect not in ['en', 'es', 'zh-cn', 'zh-tw']:
+        return jsonify({'text': "Language not found", 'entities': [], 'notes': "Language not supported."})
+
+    english = ((lang_detect == 'en') and (language == "eng"))
+    spanish = ((lang_detect == 'es') and (language == "spa"))
+    chinese = ((lang_detect[0:2] == "zh") and (language == "cmn"))
+
+    if not (english or spanish or chinese):
+        selected = "Chinese"
+        if lang_detect == "en":
+            selected = "English"
+            language = "eng"
+        elif lang_detect == "es":
+            selected = "Spanish"
+            language = "spa"
+
+        notes = "Your selected language does not seem to match the language of the text." \
+                " We will be assuming that the text is in " + selected + "."
+
+    # =====================================================================================
+    # Stanford CoreNLP
+    # =====================================================================================
+
+    nlp = StanfordCoreNLP('http://localhost:9000')
```

Comment:
this should come from command-line argument. default port is 9000, easy to be taken by others.